File: backend/app/engine/fields.py
"""Config-field rules shared by the executor: visibility (showIf) and required checks.

Mirrored by frontend/src/lib/fields.ts — keep the semantics in sync.
"""

from typing import Any

from app.engine.registry import NodeSpec


def _defaults(spec: NodeSpec) -> dict[str, Any]:
    return {f["key"]: f.get("default") for f in spec.config_fields}


def field_visible(field: dict, config: dict, defaults: dict[str, Any]) -> bool:
    show_if = field.get("showIf")
    if not show_if:
        return True
    for key, expected in show_if.items():
        actual = config.get(key)
        if actual in (None, ""):
            actual = defaults.get(key)
        # lowercase both sides so booleans compare consistently with the JS mirror
        actual_str = str(actual).lower()
        if isinstance(expected, list):
            if actual_str not in [str(v).lower() for v in expected]:
                return False
        elif actual_str != str(expected).lower():
            return False
    return True


def missing_required(spec: NodeSpec, config: dict) -> list[str]:
    """Labels of required, currently-visible fields that have no value."""
    defaults = _defaults(spec)
    missing = []
    for field in spec.config_fields:
        if not field.get("required"):
            continue
        if not field_visible(field, config, defaults):
            continue
        value = config.get(field["key"], field.get("default"))
        if value is None or (isinstance(value, str) and not value.strip()):
            missing.append(field.get("label", field["key"]))
    return missing

File: frontend/src/lib/fields.ts
/** Config-field rules: visibility (showIf) and required checks.
    Mirrors backend/app/engine/fields.py — keep the semantics in sync. */

import type { ConfigField, NodeSpec } from '../types'

export function fieldDefaults(spec: NodeSpec): Record<string, unknown> {
  const defaults: Record<string, unknown> = {}
  for (const field of spec.config_fields) defaults[field.key] = field.default
  return defaults
}

export function fieldVisible(
  field: ConfigField,
  config: Record<string, unknown>,
  defaults: Record<string, unknown>,
): boolean {
  if (!field.showIf) return true
  return Object.entries(field.showIf).every(([key, expected]) => {
    let actual = config[key]
    if (actual === undefined || actual === null || actual === '') actual = defaults[key]
    if (Array.isArray(expected)) return expected.map(String).includes(String(actual))
    return String(actual) === String(expected)
  })
}

/** Labels of required, currently-visible fields that have no value. */
export function missingRequired(spec: NodeSpec, config: Record<string, unknown>): string[] {
  const defaults = fieldDefaults(spec)
  const missing: string[] = []
  for (const field of spec.config_fields) {
    if (!field.required) continue
    if (!fieldVisible(field, config, defaults)) continue
    const value = config[field.key] ?? field.default
    if (value === undefined || value === null || (typeof value === 'string' && !value.trim())) {
      missing.push(field.label ?? field.key)
    }
  }
  return missing
}
